### PR TITLE
initial version of zreducers

### PR DESCRIPTION
**Pull request by DerThorsten**

### CODE REVIEW

**Comment by DerThorsten** [comment]
Todos:
 * reducers which have a result type which differs from the input do not work yet.
  @JohanMabille  I might need your help here to register all the needed types.
 * add all the norms ..
 * re-enable the slow unit tests...

**Comment by DerThorsten** [comment]
```
    template<class F>
    template <class T>
    inline std::size_t zreducer_functor<F>::index
    (
        const ztyped_array<T>& input_array,
        const zreducer_options& options
    )
    {
        std::size_t result;
        options.visit_reducer_options([&result, &input_array](auto&&... reduce_args)
        {
            using expr_type = decltype(F::run(input_array.get_array(), std::forward<decltype(reduce_args)>(reduce_args)...));
            using expr_value_type = typename std::decay_t<expr_type>::value_type;
            result = ztyped_array<expr_value_type>::get_class_static_index();
        });
        return result;
    }
```
gcc 4 fails to capture `input_array`.   I disable gcc 4 in the ci (for the moment,..)

File: include/zarray/zarray_impl_register.hpp [hunk 1]
```diff
@@ -0,0 +1,93 @@
+#ifndef XTENSOR_ZARRY_IMPL_REGISTER_HPP
```
Comment (JohanMabille) [inline include/zarray/zarray_impl_register.hpp:1]: Typo: should be `XTENSOR_ZARRAY_IMPL_REGISTER_HPP`

File: include/zarray/zinit.hpp [hunk 2]
```diff
@@ -0,0 +1,262 @@
...
+#include "zmath.hpp"
+#include "zreducers.hpp"
+
+
+namespace xt
+{
+    namespace mpl = xtl::mpl;
+
+  
+
```
Comment (JohanMabille) [inline include/zarray/zinit.hpp:26]: nitpicking: no need for extra new lines

File: include/zarray/zinit.hpp [hunk 3]
```diff
@@ -0,0 +1,262 @@
...
+    // static variable and be automatically
+    // called when loading a shared library
+    // for instance.
+
+    int init_zsystem();
+
+
+    /*************************************
+     * ztriple_dispatcher implementation *
+     *************************************/
```
Comment (JohanMabille) [inline include/zarray/zinit.hpp:43]: Nitpicking: this comment can be removed.

File: include/zarray/zmath.hpp [hunk 4]
```diff
@@ -270,11 +198,18 @@ namespace xt
...
 #undef XTENSOR_BINARY_ZFUNCTOR
 #undef XTENSOR_UNARY_ZFUNCTOR
 #undef XTENSOR_BINARY_ZOPERATOR
 #undef XTENSOR_UNARY_ZOPERATOR
 #undef XTENSOR_ZMAPPED_FUNCTOR
+#undef XTENSOR_ZREDUCER_FUNCTOR_HELPER
+
+
+
+
```
Comment (JohanMabille) [inline include/zarray/zmath.hpp:212]: Nitpicking:; no need for extra new lines

File: include/zarray/zreducer.hpp [hunk 5]
```diff
@@ -0,0 +1,232 @@
...
+
+        // do the actual dispatching (ie this will at some point call the actual xreducer)
+        dispatcher_type::dispatch(m_e.get_implementation(), res, args, m_reducer_options);
+        return res;
+    }
+
+
+    // this has a great overlap with xt::detail::shape_computation in xreducer
+    template <class F, class CT>
+    void zreducer<F,CT>::init_result_shape(){
```
Comment (JohanMabille) [inline include/zarray/zreducer.hpp:159]: Nitpicking: opening curly brace goes on a new line

File: include/zarray/zreducer.hpp [hunk 6]
```diff
@@ -0,0 +1,232 @@
...
+        {
+
+            std::copy(m_e.shape().cbegin(), m_e.shape().cend(), m_shape.begin());
+            for(auto a : m_reducer_options.axes())
+            {
+                m_shape[a] = 1;
+            }
+
+        }
+        else{
```
Comment (JohanMabille) [inline include/zarray/zreducer.hpp:191]: same here

File: include/zarray/zreducer.hpp [hunk 7]
```diff
@@ -0,0 +1,232 @@
...
+    auto make_zreducer(E && e, const zreducer_options & options)
+    {
+        zarray a(std::forward<E>(e));
+        using closure_type = xtl::const_closure_type_t<zarray>;
+        using reducer_type = zreducer<F, closure_type>;
+        return reducer_type(std::move(a), options);
+    }
+
+
+
```
Comment (JohanMabille) [inline include/zarray/zreducer.hpp:230]: No need for extra new lines here.

File: include/zarray/zreducer_options.hpp [hunk 8]
```diff
@@ -0,0 +1,266 @@
...
+
+namespace xt
+{
+    template <class CTE>
+    class zscalar_wrapper;
+
+    template <class F, class CT>
+    class zreducer;
+
+    class zreducer_options{
```
Comment (JohanMabille) [inline include/zarray/zreducer_options.hpp:17]: Nitpicking; opening curly brace goe son a new line

File: include/zarray/zreducer_options.hpp [hunk 9]
```diff
@@ -0,0 +1,266 @@
...
+        struct initial_tester<const xinitial<X>> : std::true_type {};
+
+
+
+        template <class F, class CT>
+        friend class zreducer;
+
+    private:
+        template<class T>
+        using make_scalar_wrapper_t = zscalar_wrapper<xscalar<T>>;
```
Comment (JohanMabille) [inline include/zarray/zreducer_options.hpp:38]: Naming; usually, `make_xxx` are used for builder functions, I think `scalar_wrapper_type` would be fine here.

File: include/zarray/zreducer_options.hpp [hunk 10]
```diff
@@ -0,0 +1,266 @@
...
+
+        bool keep_dims() const;
+
+        const auto & axes() const;
+
+        bool is_lazy()const;
+
+
+        template<class T, class F>
+        void visit_reducer_options(const bool force_lazy, F && f)const;
```
Comment (JohanMabille) [inline include/zarray/zreducer_options.hpp:115]: No need to add const for arguments passed by copy.

File: test/test_zfunction.cpp [hunk 11]
```diff
@@ -145,5 +146,21 @@ namespace xt
...
+    //     xarray<int> a = {{1, 2}, {3, 4}};
+    //     xarray<int> b = {1, 2};
+    //     xarray<int> expected = a + b;
+
+    //     zarray za(a);
+    //     zarray zb(b);
+    //     zarray zres = za + zb;
+
+    //     EXPECT_EQ(zres.get_array<int>(), expected);
+    // }
```
Comment (JohanMabille) [inline test/test_zfunction.cpp:164]: Why keeping commented code here?

File: include/zarray/zreducer_options.hpp [hunk 12]
```diff
@@ -0,0 +1,266 @@
...
+        void run(F && f, O && options) const;
+
+        svector<std::size_t> m_axes;
+        bool m_keep_dims;
+        bool m_lazy;
+
+        // it would be nice to replace this with an actual zarray itself
+        // this is not possible since this would lead to cyclic
+        // import / incomplete types.
+        std::shared_ptr<zarray_impl> m_inital_value;
```
Comment (JohanMabille) [inline include/zarray/zreducer_options.hpp:139]: typo: `m_initial_value`.

I still have a concern with the usage of `shared_ptr` here, it should not be required. What is the cyclic dependency if you include `zarray_zarray.hpp` here?

**Review by JohanMabille — CHANGES_REQUESTED** [review]
The implementation is really neat, I only have nitpicking remarks and a concern regarding the usage of the `shared_ptr` in zreducer_options. I finally think having the init functions in a dedicated header is more intuitive for the user.

Also do not forget to add the ne header to the CmakeLists.txt, otherwise they won't be installed when running `cmake install`.

**Review by JohanMabille — APPROVED** [review]

**Pull request merged**